purchase_request/models/purchase_request.py

```diff
@@ -331,3 +340,14 @@ def write(self, vals):
             requests = self.mapped('request_id')
             requests.check_auto_reject()
         return res
+
+    @api.multi
+    def unlink(self):
+        if self.filtered(
+            lambda r: r.request_state != 'draft'
+        ):
+            raise UserError(_(
+                'You can only delete a purchase request line '
+                'if the purchase request is in draft state.'
+            ))
```

Comment:
Why not use a for loop with raise such as you did in purchase_request? No need to filter everything if the first one is already in confirmed state.